### PR TITLE
worker/storageprovisioner: filesystem ops

### PR DESCRIPTION
**Pull request by axw**
This is a reworking of the filesystem code in
worker/storageprovisioner, same as what was
done recently for volumes. We now maintain a
set of "incomplete parameters" for filesystems
and filesystem attachments, and when they're
completed, schedule operations. These operations
will be attempted and rescheduled on failure,
and filesystem status will be updated according
to the outcome.

Tests have not been added for retrying or
status updates, as this branch is already huge.
I have added a card to follow up on this.

### CODE REVIEW

File: worker/storageprovisioner/blockdevices.go [hunk 1]
```diff
@@ -15,11 +15,18 @@ import (
...
-	if len(ctx.pendingFilesystems) == 0 {
+	if len(ctx.incompleteFilesystemParams) == 0 {
 		return nil
 	}
-	volumeTags := make([]names.VolumeTag, 0, len(ctx.pendingFilesystems))
-	for _, params := range ctx.pendingFilesystems {
+	volumeTags := make([]names.VolumeTag, 0, len(ctx.incompleteFilesystemParams))
+	// We only need to query volumes for incomplete filesystems,
+	// and not incomplete filesystem attachments, because a
+	// filesystem attachment cannot exist with a filesystem.
```
Comment (wallyworld) [inline worker/storageprovisioner/blockdevices.go]: s/with/without ?

Comment (axw) [inline worker/storageprovisioner/blockdevices.go]: indeed.

File: worker/storageprovisioner/filesystem_events.go [hunk 2]
```diff
@@ -0,0 +1,486 @@
...
+func filesystemAttachmentsChanged(ctx *context, ids []params.MachineStorageId) error {
+	alive, dying, dead, err := attachmentLife(ctx, ids)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	logger.Debugf("filesystem attachment alive: %v, dying: %v, dead: %v", alive, dying, dead)
+	if len(dead) != 0 {
+		// We should not see dead filesystem attachments;
+		// attachments go directly from Dying to removed.
+		logger.Debugf("unexpected dead filesystem attachments: %v", dead)
```
Comment (wallyworld) [inline worker/storageprovisioner/filesystem_events.go]: is this worth a warning?

**Comment by wallyworld** [comment]
LGTM. Please test live asap, maybe after unit tests land, or better before, so that any issues can have tests done

File: worker/storageprovisioner/filesystem_events.go (hunk 2 above)
Comment (axw) [inline worker/storageprovisioner/filesystem_events.go]: done

**Comment by axw** [comment]
> LGTM. Please test live asap, maybe after unit tests land, or better before, so that any issues can have tests done

Tested live.

**Comment by axw** [comment]
$$merge$$

**Pull request merged**